### PR TITLE
feat(sdk): Add Rust SDK

### LINKED ISSUE #6756: Support Rust among our SDKs
The SDK should implement the following:

* a way to H.recordException following the otel exception reporting spec.
* a way to instrument Rust logging libraries. if the Rust sdk supports logs natively, it would be great to route them to as LogRecords on v1/logs, but if not, fine to support them as trace events.
as part of the exception implementation, if rust http libraries follow a common format for tracing HTTP request handlers, it would be great to wrap those as native otel traces.

We should build the SDK based on the [opentelemetry sdk](https://opentelemetry.io/docs/instrumentation/rust/). There is a [guide](https://github.com/open-telemetry/opentelemetry-rust/tree/main/examples/traceresponse) for rust http specifically, so we should follow best practices here in terms of collecting the available data.

Requirements:

* a init method that sets up the Rust instrumentation to collect console logs automatically.
* a recordException method that allows capturing errors manually (see [java sdk](https://www.highlight.io/docs/sdk/java) as an example).
* Rust HTTP instrumentation that will hook into the framework to collect errors / logs automatically, without the need for a user to manually report errors.

Some getting started pointers on our [docs](https://www.highlight.io/docs/general/company/open-source/contributing/adding-an-sdk)! Please let us know if you have feedback, I'm happy to [add to them](https://github.com/highlight/highlight.io/blob/main/docs-content/general/3_company/open-source/contributing/adding-an-sdk.md).

### PR DESCRIPTION
**Pull request by mogery**
## Summary
Fixes #6756
/claim #6756

This PR adds two new Rust SDK crates, `highlightio` and `highlightio-actix`, for capturing logs and errors & instrumenting Rust code. These depend heavily on the various `opentelemetry` libraries.

## To-do
 - [x] write bare library with no framework support
 - [x] write examples for bare library
 - [x] write library for integrating into actix
 - [x] write examples for actix library
 - [x] write docs, finalise v1 crate structure

## Known issues
 - [x] ~~logging is not shown on stdout, only sent to the server~~

## How did you test this change?
I wrote examples in `sdk/highlight-rust/*/examples` and validated them using my highlight.io profile.

## Are there any deployment considerations?
You need to publish the crates in `sdk/highlight-rust`.

### CODE REVIEW

**Comment by mogery** [comment]
This is a WIP, no need to review yet, I'll un-draft the PR when it's ready :)

**Comment by mogery** [comment]
@Vadman97 @ccschmitz This is ready to go :) I mainly focused on writing docs for the highlight site, so our docs.rs site might not look the best, but that is fine in my opinion.

**Comment by mogery** [comment]
Fixed up the fails on this one, as well as the doc coherency/completeness. Everything should pass now.

**Comment by Vadman97** [comment]
Thanks @mogery ! Will be giving this a thorough review before the end of the week.

**Comment by mogery** [comment]
> Will be giving this a thorough review before the end of the week.

Alright, thanks!

**Comment by mogery** [comment]
yup! will go through all this tomorrow.

**Comment by mogery** [comment]
I'm seeing a regression of a Tokio-related bug I thought I'd fixed. Trying to get that done today, will send the screen recording after.

File: sdk/highlight-rust/Cargo.toml
[the remark this answers is not in the document]
Comment (mogery) [reply sdk/highlight-rust/Cargo.toml:1]: I'll do some dry-run crate publish tests and look into `cargo publish` CI actions and get back to you on this.

File: sdk/highlight-rust/highlightio/examples/tokio-current-thread/README.md
[the remark this answers is not in the document]
Comment (mogery) [reply sdk/highlight-rust/highlightio/examples/tokio-current-thread/README.md:1]: I've added some instructions, as well as a README for the parent directory. Is it better now?

**Comment by ccschmitz** [comment]
Removed myself as reviewer since @Vadman97 is already taking a look. @mogery, thank you for picking this up - really excited to see this!

File: sdk/highlight-rust/highlightio/examples/tokio-current-thread/README.md
[the remark this answers is not in the document]
Comment (Vadman97) [reply sdk/highlight-rust/highlightio/examples/tokio-current-thread/README.md:1]: thanks, looks good!

**Comment by Vadman97** [comment]
how's this going @mogery ? ready for another pass?

**Comment by mogery** [comment]
> how's this going @mogery ? ready for another pass?

Hi there! Sorry for the long wait. The bug that regressed is particularly obtuse and I haven't been able to make the time to try and debug it. I'll get it done by Monday.

**Comment by mogery** [comment]
@Vadman97 ready for next pass! should have some infos about crate publishing for you shortly as well.

**Comment by mogery** [comment]
Here's the screencap of the examples: https://www.youtube.com/watch?v=5svbbzBawEk

**Review by Vadman97 — APPROVED** [review]
looks great, thanks for all the hard work @mogery, and for the thorough testing in the youtube video!

would appreciate your thoughts on how we should publish the packages as rust cargo. once we have that figured out, this will be ready to merge!

**Comment by mogery** [comment]
> looks great, thanks for all the hard work @mogery, and for the thorough testing in the youtube video!
> 
> 
> 
> would appreciate your thoughts on how we should publish the packages as rust cargo. once we have that figured out, this will be ready to merge!
> 
> 

awesome! been having some super busy days. will get the cargo info to you on tuesday though. are we looking for a CI automated option or just manual is fine?

**Comment by Vadman97** [comment]
friendly bump here @mogery

**Comment by mogery** [comment]
sorry @Vadman97! been super busy lately. i've got another thing's deadline tonight, will 100% get on this tomorrow.

**Comment by mogery** [comment]
Seems like we'll need to publish the crate manually first, and then we can use [this action](https://github.com/marketplace/actions/publish-crates) to automatically detect changes and publish new versions. Could we coordinate this over Discord or some other real-time chat?

**Comment by Vadman97** [comment]
another bump here @mogery . if you think this is in a close spot, i can pull the CI work into a new ticket

**Comment by mogery** [comment]
> another bump here @mogery . if you think this is in a close spot, i can pull the CI work into a new ticket

I can work on this now, I posted a comment with info 4 hours ago.

**Comment by mogery** [comment]
Added the READMEs and the GitHub Action. Not sure how to make the action actually run in this PR though. Maybe it only runs once merged.

**Comment by Vadman97** [comment]
> Run katyo/publish-crates@v2
> Searching cargo packages at '.'
> /home/runner/.cargo/bin/cargo metadata --no-deps --format-version [1](https://github.com/highlight/highlight/actions/runs/8179431922/job/22366690967?pr=7578#step:4:1) --manifest-path Cargo.toml
error: manifest path `Cargo.toml` does not exist
Error: Error: The process '/home/runner/.cargo/bin/cargo' failed with exit code 101

looks like the current working directory isn't set right?

**Comment by mogery** [comment]
> > Run katyo/publish-crates@v2
> > Searching cargo packages at '.'
> > /home/runner/.cargo/bin/cargo metadata --no-deps --format-version [1](https://github.com/highlight/highlight/actions/runs/8179431922/job/22366690967?pr=7578#step:4:1) --manifest-path Cargo.toml
> > error: manifest path `Cargo.toml` does not exist
> > Error: Error: The process '/home/runner/.cargo/bin/cargo' failed with exit code 101
> 
> looks like the current working directory isn't set right?

oops! pushed a fix 🤞

File: .github/workflows/rust-sdk.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            - name: Build highlightio-actix
+              uses: katyo/publish-crates@v2
+              with:
+                  path: ./sdk/highlight-rust/
+                  dry-run: true
+                  args: -p highlightio-actix
+    publish:
+        name: Publish
+        runs-on: ubuntu-latest
+        if: github.event_name == 'push'
```
Comment (Vadman97) [inline .github/workflows/rust-sdk.yml:56]: awesome, so the dry run will happen in branches, but the cargo publish will only actually happen on main push.

**Review by Vadman97 — APPROVED** [review]
sweet! thanks for getting the CI done as part of this too 🙏

**Comment by SpennyNDaJets** [comment]
@all-contributors please add @mogery for code

**Pull request merged**